patches/0037-Non-idempotent-sns-topic-creation.patch

```diff
@@ -0,0 +1,64 @@
+diff --git a/internal/service/sns/topic.go b/internal/service/sns/topic.go
+index 98ebd87d5d..adc13a9f58 100644
+--- a/internal/service/sns/topic.go
++++ b/internal/service/sns/topic.go
+@@ -9,6 +9,7 @@ import (
+ 	"fmt"
+ 	"log"
+ 	"regexp"
++	"sync"
+ 	"time"
+ 
+ 	"github.com/YakDriver/regexache"
+@@ -259,6 +260,8 @@ func resourceTopic() *schema.Resource {
+ 	}
+ }
+ 
++var snsGlobalMutex sync.Mutex
++
+ func resourceTopicCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+ 	var diags diag.Diagnostics
+ 	conn := meta.(*conns.AWSClient).SNSClient(ctx)
+@@ -283,6 +286,34 @@ func resourceTopicCreate(ctx context.Context, d *schema.ResourceData, meta inter
+ 		delete(attributes, topicAttributeNameFIFOTopic)
+ 	}
+ 
++	// Look up if the topic already exists
++	snsGlobalMutex.Lock()
++	var next_token *string = nil
++	for done := false; !done; {
++		output, err := conn.ListTopics(ctx, &sns.ListTopicsInput{
```

Comment:
Looping over ALL the topics is not going to be acceptable, is there a way to check one of them only? Perhaps this file has some resourceTopicRead or some such method as provider typically needs to handle importing resources from the cloud and handling not found.

Reply:
Yeah copilot thinks so too:

```
func constuctTopicArn(client sns.Client, account, region, snsTopicName string) string {
	return fmt.Sprintf("arn:aws:sns:%s:%s:%s", region, account, snsTopicName)
}
```

I think the code under resourceTopicRead is indeed addressing some obvious concerns:

```
	attributes, err := findTopicAttributesWithValidAWSPrincipalsByARN(ctx, conn, d.Id())

	if !d.IsNewResource() && tfresource.NotFound(err) {
		log.Printf("[WARN] SNS Topic (%s) not found, removing from state", d.Id())
		d.SetId("")
		return nil
	}
```

It's using one GetAttributes call instead of listing potentially many topics, and it retries the call for network failures thus making it more robust.